pkg/frontend/openshiftcluster_preflightvalidation.go

```diff
@@ -0,0 +1,109 @@
+	originalPath := r.Context().Value(middleware.ContextKeyOriginalPath).(string)
+	referer := r.Header.Get("Referer")
+
+	subId := chi.URLParam(r, "subscriptionId")
+	resourceProviderNamespace := chi.URLParam(r, "resourceProviderNamespace")
+
+	apiVersion := r.URL.Query().Get(api.APIVersionKey)
+	err := cosmosdb.RetryOnPreconditionFailed(func() error {
+		var err error
+		b, err = f._preflightValidation(ctx, log, body, correlationData, systemData, r.URL.Path, originalPath, r.Method, referer, &header, f.apis[apiVersion].OpenShiftClusterConverter, f.apis[apiVersion].OpenShiftClusterStaticValidator, subId, resourceProviderNamespace, apiVersion)
+		return err
+	})
+
+	frontendOperationResultLog(log, r.Method, err)
+	reply(log, w, header, b, err)
+}
+
+func (f *frontend) _preflightValidation(ctx context.Context, log *logrus.Entry, body []byte, correlationData *api.CorrelationData, systemData *api.SystemData, path, originalPath, method, referer string, header *http.Header, converter api.OpenShiftClusterConverter, staticValidator api.OpenShiftClusterStaticValidator, subId, resourceProviderNamespace string, apiVersion string) ([]byte, error) {
+	// multiple errors should be able to return, just add them to the message
+	/*subscription, err := f.validateSubscriptionState(ctx, path, api.SubscriptionStateRegistered)
+	if err != nil {
+		return nil, err
+	}
+	*/
+
+	doc, err := f.dbOpenShiftClusters.Get(ctx, path)
+	if err != nil && !cosmosdb.IsErrorStatusCode(err, http.StatusNotFound) {
+		return nil, err
+	}
+
+	newProvisionState := doc == nil
+
+	// dont create this as it would require creating openshiftclusterdoc
+	// instead grab info you may need like location that is given to the doc
+	/*
+			originalR, err := azure.ParseResourceID(originalPath)
+		if err != nil {
+			return nil, err
+		}
+
+		doc = &api.OpenShiftClusterDocument{
+			ID:  f.dbOpenShiftClusters.NewUUID(),
+			Key: path,
+			OpenShiftCluster: &api.OpenShiftCluster{
+				ID:   originalPath,
+				Name: originalR.ResourceName,
+				Type: originalR.Provider + "/" + originalR.ResourceType,
+				Properties: api.OpenShiftClusterProperties{
+					ArchitectureVersion: version.InstallArchitectureVersion,
+					ProvisioningState:   api.ProvisioningStateSucceeded,
+					CreatedAt:           f.now().UTC(),
+					CreatedBy:           version.GitCommit,
+					ProvisionedBy:       version.GitCommit,
+				},
+			},
+		}
+		if !f.env.IsLocalDevelopmentMode() {
+			doc.OpenShiftCluster.Properties.FeatureProfile.GatewayEnabled = true
+		}
+	*/
+
+	if !newProvisionState {
+		err = validateTerminalProvisioningState(doc.OpenShiftCluster.Properties.ProvisioningState)
+		if err != nil {
+			return nil, err
+		}
+	} else {
+		// validateTerminalProvisionState which if creating, wouldn't have. maybe save that check for deploying
+
+		// preflight doc mentions SKU and quota validation are good candidates to do preflight validation on
```

Comment:
More things this code could check:
- Ensure VNet has no NSGs attached. I've gotten this error before.
- Ensure service principal client ID and secret aren't empty. Customers have done this before.
- Verify subnet CIDR ranges. Since we attach a load balancer, we need to have enough IP addresses.

Reply:
Let me know if you're adding the NSG check as I'm working on BYO NSG.  That would allow pre-attached NSGs on the subnet.